pages/blog/typescript-frontend-backend.mdx

```diff
@@ -0,0 +1,127 @@
+
+The next step is to extract the `run()` method’s return type to get the `type` of the API response
+
+We can use a helper like [`type-fest`’s](https://www.npmjs.com/package/type-fest) `PromiseValue` to get the return value, or we can do it ourselves:
+
+```ts
+// from https://www.jpwilliams.dev/how-to-unpack-the-return-type-of-a-promise-in-typescript
+
+export type UnwrapPromise<T> = T extends Promise<infer U>
+  ? U
+  : T extends (...args: any) => Promise<infer U>
+  ? U
+  : T extends (...args: any) => infer U
+  ? U
+  : T;
+```
+
+So, the type of the Action’s response is:
+
+```ts
+type ActionResponse = UnwrapPromise<typeof GetTime.prototype.run>; // = { time: number; }
+```
+
+And in our IDE:
+
+<Image
+  width={"100%"}
+  alt="Share Dataset"
+  src={require("../../public/posts/typescript-frontend-backend/screenshot-1.png")}
+/>
+
+This is excellent because now any changes to our action will result in the `type` being automatically updated!
+
+## Consuming the API Response Type in React
+
+The Grouparoo Application is stored in a [monorepo](https://github.com/grouparoo/grouparoo), which means that the frontend and backend code always exist side-by-side. This means that we can reference the API code from our Frontend code, and make a helper to check our response types. We don't need our API code at run-time, but we can import the `types` from it as we develop and compile the app to Javascript.
+
+The first thing to do is make a utility file which imports our Actions and extracts their types. Grouparoo does this in `web/utils/apiData.ts`
+
+```ts
+import { UnwrapPromise } from "./UnwrapPromise";
+import { GetTime } from "../../api/src/actions/getTime";
+
+export namespace Actions {
+  export type GetTime = UnwrapPromise<typeof GetTime.prototype.run>;
+}
+```
+
+This `apiData.ts` will allow us to more concisely reference `Actions.GetTime` in the rest of our react application.
+
+Now, to use the Action’s response type, all we have to do is assign it to the response of an API request:
+
+```ts
+import { useState, useEffect } from "react";
+import { Actions } from "../utils/apiData";
+
+export default function TimeComponent() {
+  const [time, setTime] = useState(0);
+
+  useEffect(() => {
+    load();
+  }, []);
+
+  async function load() {
+    const response: Actions.GetTime = await fetch("/api/time");
+    setTime(response.time);
+  }
+
+  if (time === 0) return <div>loading...</div>;
+
+  const formattedTime = new Date(time).toLocaleString();
+  return <div>The time is: {formattedTime}</div>;
+}
+```
+
+Now we have enforced that the type of `response` in the `load()` method above will match the Action, being `{ time: number; }`. We will now get help from Typescript if don’t use that response value as a number, if we try to assign it to a string variable or otherwise use it improperly.
```

Comment:
"We will now get help from Typescript if don’t use that response value as a number, if we try to assign it to a string variable or otherwise use it improperly." 

Too many "if"s here? Can you take another pass at this?